src/main/core/worker.rs

```diff
@@ -1,188 +1,177 @@
-        let memory_manager =
-            unsafe { &mut *(cshadow::process_getMemoryManager(cprocess) as *mut MemoryManager) };
-
+    /// Set the currently-active Host.
+    pub fn set_active_host(host: &Host) {
+        let info = host.info().clone();
+        let c_host = notnull_mut_debug(host.chost());
+        unsafe { cshadow::host_ref(c_host) };
         WORKER.with(|worker| {
-            // Once the process lives in Rust, the MemoryManager will be a
-            // member of it, so accessing it will require holding a reference to
-            // the process.
-            //
-            // It *may* turn out to be too restrictive to hold a mutable
-            // reference to the process while accessing memory, in which case
-            // the process can store a RefCell<MemoryManager>, and which we
-            // could simulate here with an immutable borrow of the process, and
-            // a mutable borrow of a stand-in RefCell for the MemoryManager.
-            let worker = worker.borrow();
-            let _borrow = worker.as_ref().unwrap().active_process.borrow_mut();
-            f(memory_manager)
-        })
+            let mut worker = worker.get().unwrap().borrow_mut();
+            let old = worker.active_host_info.replace(info);
+            debug_assert!(old.is_none());
+            debug_assert!(worker.c_active_host.is_null());
+            worker.c_active_host = notnull_mut_debug(c_host);
+        });
     }
 
-    /// Run `f` with a reference to the active process.
-    pub fn with_active_process<F, R>(f: F) -> R
-    where
-        F: FnOnce(&Process) -> R,
-    {
-        let process = unsafe {
-            let cprocess = cshadow::worker_getActiveProcess();
-            assert!(!cprocess.is_null());
-            Process::borrow_from_c(cprocess)
-        };
-        WORKER.with(|worker| {
-            let worker = worker.borrow();
-            let _borrow = worker.as_ref().unwrap().active_process.borrow();
-            f(&process)
-        })
+    /// Clear the currently-active Host.
+    pub fn clear_active_host() {
+        let c_host = WORKER.with(|worker| {
+            let mut worker = worker.get().unwrap().borrow_mut();
+            let old = worker.active_host_info.take();
+            debug_assert!(!old.is_none());
+            let old = worker.c_active_host;
+            debug_assert!(!old.is_null());
+            worker.c_active_host = std::ptr::null_mut();
+            old
+        });
+        unsafe { cshadow::host_unref(c_host) };
     }
 
-    /// Run `f` with a reference to the current process.
-    pub fn with_active_process_mut<F, R>(f: F) -> R
-    where
-        F: FnOnce(&mut Process) -> R,
-    {
-        let mut process = unsafe {
-            let cprocess = cshadow::worker_getActiveProcess();
-            assert!(!cprocess.is_null());
-            Process::borrow_from_c(cprocess)
+    /// Set the currently-active Process.
+    pub fn set_active_process(process: &Process) {
```

Comment:
If the active host is already set, should there be checks to make sure that this process belongs to that host (and same for the thread)?

Reply:
Done